bridges/snowbridge/pallets/system-frontend/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+#[cfg(feature = "runtime-benchmarks")]
+pub trait BenchmarkHelper<O>
+where
+	O: OriginTrait,
+{
+	fn make_xcm_origin(location: Location) -> O;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Origin check for XCM locations that can create agents
+		type CreateAgentOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
+
+		/// Origin check for XCM locations that can create agents
+		type RegisterTokenOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
+
+		/// XCM message sender
+		type XcmSender: SendXcm;
+
+		/// To withdraw and deposit an asset.
+		type AssetTransactor: TransactAsset;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		type Helper: BenchmarkHelper<Self::RuntimeOrigin>;
+
+		type FeeAsset: Get<Location>;
+
+		type DeliveryFee: Get<Asset>;
+
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A `CreateAgent` message was sent to Bridge Hub
+		CreateAgent { location: Location },
+		/// A message to register a Polkadot-native token was sent to Bridge Hub
+		RegisterToken {
+			/// Location of Polkadot-native token
+			location: Location,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Convert versioned location failure
+		UnsupportedLocationVersion,
+		/// Check location failure, should start from the dispatch origin as owner
+		OwnerCheck,
+		/// Send xcm message failure
+		Send,
+		/// Withdraw fee asset failure
+		FundsUnavailable,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Call `create_agent` to instantiate a new agent contract representing `origin`.
+		/// - `origin`: Can be any sovereign `Location`
+		/// - `fee`: Fee in Ether paying for the execution cost on Ethreum
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::create_agent())]
+		pub fn create_agent(origin: OriginFor<T>, fee: u128) -> DispatchResult {
+			let origin_location = T::CreateAgentOrigin::ensure_origin(origin)?;
+
+			Self::burn_fees(origin_location.clone(), fee)?;
+
+			let call = SnowbridgeControl::Control(ControlCall::CreateAgent {
+				location: Box::new(VersionedLocation::from(origin_location.clone())),
+				fee,
+			});
+
+			let xcm: Xcm<()> = vec![
+				ReceiveTeleportedAsset(T::DeliveryFee::get().into()),
+				PayFees { asset: T::DeliveryFee::get() },
+				Transact {
+					origin_kind: OriginKind::Xcm,
+					call: call.encode().into(),
+					fallback_max_weight: None,
+				},
+			]
+			.into();
+
+			Self::send(xcm)?;
+
+			Self::deposit_event(Event::<T>::CreateAgent { location: origin_location.clone() });
+			Ok(())
+		}
+
+		/// Registers a Polkadot-native token as a wrapped ERC20 token on Ethereum.
+		/// - `origin`: Must be `Location` from a sibling parachain
```

Comment:
@yrong What about DOT, which comes from the relay chain rather than a sibling parachain? Should we amend the documentation?

Reply:
I'd assume DOT can be registered from AH by anyone.